models/src/main/scala/coop/rchain/models/PrettyPrinter.scala

```diff
@@ -9,220 +9,180 @@ import coop.rchain.models.Var.VarInstance
-      case v: Var     => prettyPrint(v)
-      case c: Channel => prettyPrint(c)
+      case v: Var     => buildString(v)
+      case c: Channel => buildString(c)
       case s: Send =>
-        prettyPrint(s.chan.get)
-        print("!(")
-        for { (datum, i) <- s.data.zipWithIndex } yield {
-          prettyPrint(datum)
-          if (i != s.data.length - 1)
-            print(",")
-        }
-        print(")")
+        buildString(s.chan.get) + {
+          if (s.persistent) "!!("
+          else "!("
+        } + {
+          ("" /: s.data.zipWithIndex) {
+            case (string, (datum, i)) =>
+              string + buildString(datum) + {
+                if (i != s.data.length - 1) ", "
+                else ""
+              }
+          }
+        } + ")"
+
       case r: Receive =>
-        print("for (")
-        for { (bind, i) <- r.binds.zipWithIndex } yield {
-          printSeq(bind.patterns)
-          print(" <- ")
-          prettyPrint(bind.source.get)
-          if (i != r.binds.length - 1) {
-            print("; ")
+        "for( " + {
+          ("" /: r.binds.zipWithIndex) {
+            case (string, (bind, i)) =>
+              // Stepping past a channel increments boundShift by 1.
+              // Entering a pattern sets freeShift = boundShift and boundShift = 0.
+              string + PrettyPrinter(freeShift, boundShift + i).buildPattern(bind.patterns) + {
+                if (r.persistent) " <= " + buildString(bind.source.get)
+                else " <- " + buildString(bind.source.get)
+              } + {
+                if (i != r.binds.length - 1) " ; "
+                else ""
+              }
           }
-        }
-        print(") { ")
-        prettyPrint(r.body.get)
-        print(" }")
-      case e: Eval =>
-        print("*")
-        prettyPrint(e.channel.get)
+        } + " ) { " + PrettyPrinter(freeShift, boundShift + r.binds.length)
+          .buildString(r.body.get) + " }"
+
+      case e: Eval => "*" + buildString(e.channel.get)
       case n: New =>
-        print("new ")
-        printNewVariables(n.bindCount)
-        print(" in {")
-        prettyPrint(n.p.get)
-        print("}")
+        "new " + buildNewVariables(n.bindCount) + " in { " + PrettyPrinter(
+          freeShift,
+          boundShift + n.bindCount).buildString(n.p.get) + " }"
       case e: Expr =>
-        prettyPrint(e)
+        buildString(e)
       case m: Match =>
-        print("match { ")
-        prettyPrint(m.target.get)
-        print(" } { ")
-        for { (matchCase, i) <- m.cases.zipWithIndex } yield {
-          prettyPrint(matchCase.pattern.get)
-          print(" => ")
-          prettyPrint(matchCase.source.get)
-          if (i != m.cases.length - 1) {
-            print("; ")
+        "match { " + buildString(m.target.get) + " } { " + {
+          for { (matchCase, i) <- m.cases.zipWithIndex } yield {
+            buildString(matchCase.pattern.get) + " => " + buildString(matchCase.source.get) + {
```

Comment:
matchCase.pattern is a pattern, and so it should be processed with freeShift = bindShift, bindShift = 0.

On the same line, the body should have bindShift = bindShift + matchCase.pattern.get.freeCount